### ISSUE TITLE
#0 failed to flush the buffer. error_class=TypeError error="no implicit conversion of nil into String"

### ISSUE THREAD

**Issue by chs-bnet** [uken/fluent-plugin-elasticsearch#303]
#### Problem
We get this error, including a traceback in the logs. Once this error appears, the fluentd elasticsearch plugin no longer sends ANY logs to elasticsearch. The log in question that causes the error must be tracked down, deleted/cleared and td-agent restarted before logs will flow into elasticsearch again.

Error:
```
2017-10-05 21:41:07 +0000 [warn]: #0 failed to flush the buffer. retry_time=0 next_retry_seconds=2017-10-05 21:41:08 +0000 chunk="55ad3911cab502a6aa2d73d4ea3ea1a4" error_class=TypeError error="no implicit conversion of nil into String"
  2017-10-05 21:41:07 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluent-plugin-elasticsearch-2.0.0.rc.5/lib/fluent/plugin/out_elasticsearch_dynamic.rb:247:in `sub!'
  2017-10-05 21:41:07 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluent-plugin-elasticsearch-2.0.0.rc.5/lib/fluent/plugin/out_elasticsearch_dynamic.rb:247:in `expand_param'
  2017-10-05 21:41:07 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluent-plugin-elasticsearch-2.0.0.rc.5/lib/fluent/plugin/out_elasticsearch_dynamic.rb:137:in `block (2 levels) in write'
  2017-10-05 21:41:07 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluent-plugin-elasticsearch-2.0.0.rc.5/lib/fluent/plugin/out_elasticsearch_dynamic.rb:132:in `each'
  2017-10-05 21:41:07 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluent-plugin-elasticsearch-2.0.0.rc.5/lib/fluent/plugin/out_elasticsearch_dynamic.rb:132:in `each_with_index'
  2017-10-05 21:41:07 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluent-plugin-elasticsearch-2.0.0.rc.5/lib/fluent/plugin/out_elasticsearch_dynamic.rb:132:in `block in write'
  2017-10-05 21:41:07 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.14.21/lib/fluent/event.rb:323:in `each'
  2017-10-05 21:41:07 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.14.21/lib/fluent/event.rb:323:in `block in each'
  2017-10-05 21:41:07 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.14.21/lib/fluent/plugin/buffer/file_chunk.rb:170:in `open'
  2017-10-05 21:41:07 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.14.21/lib/fluent/event.rb:322:in `each'
  2017-10-05 21:41:07 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluent-plugin-elasticsearch-2.0.0.rc.5/lib/fluent/plugin/out_elasticsearch_dynamic.rb:128:in `write'
  2017-10-05 21:41:07 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.14.21/lib/fluent/plugin/output.rb:1061:in `try_flush'
  2017-10-05 21:41:07 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.14.21/lib/fluent/plugin/output.rb:1286:in `flush_thread_run'
  2017-10-05 21:41:07 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.14.21/lib/fluent/plugin/output.rb:438:in `block (2 levels) in start'
  2017-10-05 21:41:07 +0000 [warn]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.14.21/lib/fluent/plugin_helper/thread.rb:78:in `block in thread_create'
```

#### Steps to replicate

Example of log in question:
```
# msgpack2json -di buffer.q55ad3911cab502a6aa2d73d4ea3ea1a4.log
[
    <ext of type 0 size 8>,
    {
        "message": "fluentd main process get SIGTERM",
        "underscore_sumo_metadata": {
            "category": "dev/debug/us-east-2////",
            "source": "",
            "host": "fluent",
            "log_format": "json"
        },
        "fluent_added_timestamp": "2017-10-05T21:40:02.727+0000",
        "@timestamp": "2017-10-05T21:40:02.727+0000",
        "td_agent_id": "SANITIZED",
        "td_tag": "SANITIZED",
        "ingest_region": "us-east-2"
    }
]
# msgpack2json -di buffer.q55ad3911cab502a6aa2d73d4ea3ea1a4.log.meta 
{
    "timekey": 1507237200,
    "tag": "SANITIZED",
    "variables": null,
    "id": <bin of size 16>,
    "s": 2,
    "c": 1507239602,
    "m": 1507239602
}
```
Originally, we thought it might be lack of timestamp. So we make sure to put everything through a filter that fixes and/or adds a valid timestamp. Then we found a key `_sumo_metadata` that started with an underscore and believed that might be part of it, so now we're renaming any leading `_` as `underscore_`, which you can see above.

However, with those two "fixes" in place the issue is still happening and now we're at a loss as to what might be causing it.

Our sanitized fluentd configuration:

```
<system>
  log_level info
</system>

<filter **>
  @type elasticsearch_timestamp_check
  log_level debug
</filter>

<filter **>
  @type rename_key
  rename_rule1 ^_(.+) underscore_${md[1]}
</filter>


<filter **>
  @type record_transformer
  <record>
    td_agent_id <SANITIZED>
    td_tag ${tag}
    ingest_region us-east-2
  </record>
</filter>

<source>
  @type monitor_agent
  bind 0.0.0.0
  port 24220
</source>

# Internal Forwarder
<source>
  @type forward
  port 25140
  bind 0.0.0.0
</source>

# External Forwarder
<source>
  port 2514
  @type secure_forward
  allow_keepalive false
  shared_key                    <SANITIZED>
  self_hostname                 <SANITIZED>
  secure                        true
  ca_cert_path                  /etc/td-agent/ca_cert.pem
  ca_private_key_path           /etc/td-agent/ca_key.pem
  ca_private_key_passphrase     <SANITIZED>
  buffer_type memory
  buffer_chunk_limit 2m
  buffer_queue_limit 4096
  slow_flush_log_threshold 60s
  authentication                yes

  <user>
    username <SANITIZED>
    password <SANITIZED>
  </user>
  <user>
    username <SANITIZED>
    password <SANITIZED>
  </user>
</source>

<match prefix.**>
  @type copy
    <store>
      @type elasticsearch_dynamic
      hosts <SANITIZED>
      port 9200
      slow_flush_log_threshold 300s
      logstash_format true
      logstash_prefix prefix-${tag_parts[5]}-${tag_parts[6]}

      <buffer tag,time>
        @type file
        path /data/td-agent/prefix-buffer
        timekey     60m
        timekey_wait 1m
      </buffer>

      buffer_chunk_limit 2m
      buffer_queue_limit 2048
      request_timeout 30s
      reload_on_failure true
      flush_interval 10s
      retry_wait 1s
      num_threads 1
    </store>

    <store>
      @type s3
      buffer_path /data/td-agent/s3_prefix
      num_threads 2
      slow_flush_log_threshold 120s
      flush_interval 10s
      include_time_key true
      include_tag_key true
      path logs/prefix/%Y/%m/%d
      s3_bucket <SANITIZED>
      s3_region us-east-2
      s3_object_key_format %{path}/%{index}_<SANITIZED>.%{file_extension}
      store_as gzip_command
      utc
    </store>

    <store>
      @type gelf
      host <SANITIZED>
      port 12201
      flush_interval 5s
      protocol tcp
    </store>
</match>

<match **>
  @type copy
    <store>
      @type elasticsearch_dynamic
      hosts <SANITIZED>
      port 9200
      logstash_format true
      logstash_prefix ${tag_parts[0]}-${tag_parts[1]}

      <buffer tag,time>
        @type file
        path /data/td-agent/buffer
        timekey     60m
        timekey_wait 1m
      </buffer>

      buffer_chunk_limit 2m
      buffer_queue_limit 2048
      request_timeout 30s
      reload_connections false
      reload_on_failure true
      flush_interval 10s
      retry_wait 1s
      num_threads 1
      slow_flush_log_threshold 120s
    </store>

    <store>
      @type s3
      buffer_path /data/td-agent/s3
      num_threads 2
      flush_interval 10s
      include_time_key true
      include_tag_key true
      path logs/catch-all/%Y/%m/%d
      s3_bucket <SANITIZED>
      s3_region us-east-2
      s3_object_key_format %{path}/%{index}_<SANITIZED>.%{file_extension}
      store_as gzip_command
      utc
      slow_flush_log_threshold 120s
    </store>
    <store>
      @type gelf
      host <SANITIZED>
      protocol tcp
      port 12201
      flush_interval 5s
    </store>
</match>
```

#### Using Fluentd and ES plugin versions
```
# /opt/td-agent/embedded/bin/fluent-gem list --local

*** LOCAL GEMS ***

actionmailer (4.2.8)
actionpack (4.2.8)
actionview (4.2.8)
activejob (4.2.8)
activemodel (4.2.8)
activerecord (4.2.8)
activesupport (4.2.8)
addressable (2.5.2, 2.5.1)
arel (6.0.4)
aws-sdk (2.10.45)
aws-sdk-core (2.10.45)
aws-sdk-resources (2.10.45)
aws-sigv4 (1.0.2)
bigdecimal (default: 1.2.4)
bson (4.1.1)
builder (3.2.3)
bundler (1.14.5)
celluloid (0.15.2)
cool.io (1.5.1)
crass (1.0.2)
diff-lcs (1.3)
draper (1.4.0)
elasticsearch (5.0.4)
elasticsearch-api (5.0.4)
elasticsearch-transport (5.0.4)
erubis (2.7.0)
excon (0.59.0)
faraday (0.13.1)
fluent-logger (0.7.1)
fluent-mixin-plaintextformatter (0.2.6)
fluent-plugin-elasticsearch (2.0.0.rc.5, 1.10.0)
fluent-plugin-elasticsearch-timestamp-check (0.2.5)
fluent-plugin-gelf-hs (1.0.4)
fluent-plugin-kafka (0.6.1)
fluent-plugin-mongo (0.8.1)
fluent-plugin-rename-key (0.4.1)
fluent-plugin-rewrite-tag-filter (1.5.6)
fluent-plugin-route (1.0.0)
fluent-plugin-s3 (0.8.5)
fluent-plugin-scribe (0.10.14)
fluent-plugin-secure-forward (0.4.5)
fluent-plugin-td (0.10.29)
fluent-plugin-td-monitoring (0.2.3)
fluent-plugin-webhdfs (0.7.1)
fluentd (0.14.21, 0.12.40)
fluentd-ui (0.4.4)
font-awesome-rails (4.7.0.1)
gelf (3.0.0)
globalid (0.4.0)
haml (4.0.7)
haml-rails (0.5.3)
hike (1.2.3)
hirb (0.7.3)
http_parser.rb (0.6.0)
httpclient (2.8.2.4)
i18n (0.8.1)
io-console (default: 0.4.3)
ipaddress (0.8.3)
jbuilder (2.6.3)
jmespath (1.3.1)
jquery-rails (3.1.4)
json (default: 1.8.1)
kramdown (1.13.2)
kramdown-haml (0.0.3)
loofah (2.1.1, 2.0.3)
ltsv (0.1.0)
mail (2.6.4)
mime-types (3.1)
mime-types-data (3.2016.0521)
mini_portile2 (2.3.0, 2.1.0)
minitest (5.10.1, default: 4.7.5)
mixlib-cli (1.7.0)
mixlib-config (2.2.4)
mixlib-log (1.7.1)
mixlib-shellout (2.2.7)
mongo (2.2.7)
msgpack (1.1.0)
multi_json (1.12.1)
multipart-post (2.0.0)
nokogiri (1.8.1)
ohai (6.20.0)
oj (2.18.1)
parallel (1.8.0)
plugin (1.1)
proxifier (1.0.3)
psych (default: 2.0.5)
public_suffix (3.0.0, 2.0.5)
puma (3.8.2)
rack (1.6.5)
rack-test (0.6.3)
rails (4.2.8)
rails-deprecated_sanitizer (1.0.3)
rails-dom-testing (1.0.8)
rails-html-sanitizer (1.0.3)
railties (4.2.8)
rake (default: 10.1.0)
rdoc (default: 4.1.0)
request_store (1.3.2)
resolve-hostname (0.1.0)
ruby-kafka (0.4.2)
ruby-progressbar (1.8.3)
ruby_dig (0.0.2)
rubyzip (1.2.1, 1.1.7)
sass (3.2.19)
sass-rails (4.0.5)
serverengine (2.0.5)
settingslogic (2.0.9)
sigdump (0.2.4)
sprockets (2.12.4)
sprockets-rails (2.3.3)
string-scrub (0.0.5)
strptime (0.1.9)
sucker_punch (1.0.5)
systemu (2.5.2)
td (0.15.2)
td-client (0.8.85)
td-logger (0.3.27)
test-unit (default: 2.1.10.0)
thor (0.19.4)
thread_safe (0.3.6)
thrift (0.8.0)
tilt (1.4.1)
timers (1.1.0)
tzinfo (1.2.3)
tzinfo-data (1.2017.2)
uuidtools (2.1.5)
webhdfs (0.8.0)
yajl-ruby (1.3.0)
zip-zip (0.3)
```

**Comment by cosmo0920** [comment]
Using multiple Fuentd versions is very harmful.
Could you try ES plugin 2.0.0.rc.5 and Fluentd 0.14.21 combination?
And now out_elasticsearch supports built-in placehodlers like:

```aconf
logstash_prefix prefix-${tag_parts[5]}-${tag_parts[6]}
```

This is also work `@type elasticsearch`.

**Comment by chs-bnet** [comment]
We actually figured out what the issue was. We were matching on `${tag_parts[5]}` and `[6]`, but if you look at that message it only has 3 total tag_parts. We switched to `${tag_parts[-2]}` and `[-1]` and that seems to have fixed the issue.

That said, I wanted to leave the issue open because:
* The plugin should catch this as an exception
* The plugin should have better error messages in this case
* The plugin should not crash and fail to deliver messages to Elasticsearch because of one malformed tag
